sycl/doc/extensions/experimental/sycl_ext_intel_fpga_task_sequence.asciidoc

```diff
@@ -0,0 +1,530 @@
+  float* out = sycl::malloc_host<float>(1, q);
+  ...
+
+  q.single_task([=] {
+    task_sequence<vectorSum> firstHalf;
+    task_sequence<vectorSum> secondHalf;
+    constexpr int halfCount = kCount/2;
+    firstHalf.async(in, 0, halfCount);
+    secondHalf.async(in, halfCount+1, kCount-1);
+    *out = firstHalf.get() + secondHalf.get();
+  }).wait();
+
+...
+
+```
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_INTEL_FPGA_TASK_SEQUENCE` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to determine
+if the implementation supports this feature, or applications can test the
+macro's value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+== Definitions
+
+== `task_sequence` class
+
+[source,c++,linenums]
+----
+namespace sycl::ext::intel::experimental {
+
+template <auto &f, typename PropertyListT = empty_properties_t>
+class task_sequence {
+
+public:
+  task_sequence() = default;
+
+  task_sequence(const task_sequence &) = delete;
+  task_sequence &operator=(const task_sequence &) = delete;
+  task_sequence(task_sequence &&) = delete;
+  task_sequence &operator=(task_sequence &&) = delete;
+
+
+  void async(__unspecified__... args);
+
+  __unspecified__ get();
+
+  template<typename propertyT>
+  static constexpr bool has_property();
+
+  // The return type is an unspecified internal class used to represent
+  // instances of propertyT
+  template<typename propertyT>
+  static constexpr /*unspecified*/ get_property();
+
+  ~task_sequence();
+}
+
+} // namespace sycl::ext::intel::experimental
+
+----
+
+`task_sequence` is a class template, parameterized by an `auto` reference to a
+Callable `f`, and a list of properties _PropertyListT_. The Callable `f` defines
+the asynchronous task to be associated with the `task_sequence`, and requiring
+an auto reference ensures that each `f` is statically resolvable at compile
+time. Static resolvability by the compiler is desirable when compiling for
+spatial architectures as it can enable the generation of more efficient
+hardware. Templating a task_sequence on a non-Callable is malformed.
+_PropertyListT_ enables properties to be associated with a `task_sequence`
+
+[cols="1,1"]
+|===
+|Member functions of the `task_sequence` class|Description
+
+|void async(\\__unspecified__... args)
+|Available only in device functions. +
+ +
+ Asynchronously invoke the task_sequence's associated Callable `f` with `args`
+ Calling `async` on a task_sequence with mismatched argument types with respect
+ to the Callable `f`'s argument types is malformed.
+
+|\\__unspecified__ get()
```

Comment:
C++ std::async returns a handle that you'd just call handle::get() on to block and retrieve the results. I don't think this concept works differently for a task_sequence, and storing multiple results does not affect which result gets retrieved. 
i.e. The first get() call retrieves the results of the first async() call. The second get() call retrieves the results of the second async() call, etc.

task_sequence as a construct mapped to hardware is like a unique hardware instance for each templated task function. Wherever this task_sequence is in scope, async() and get() can be called if calculation is needed on a sidepath such that it does not stall the main datapath.

Reply:
Ah okay so this is aiming to align with the C++ `std::async`, in that case I think it makes sense to keep the name as it is, and thanks for the explanation of how this is mapped to hardware.

A thought, if we're looking to map to `std::async` perhaps if we were to use a SYCL future type here it would make it clearer that `get` performs a similar function to the `std::future::get`. So `task_sequence::async` would return a `sycl::future<T>` where `T` is the return type of the function, and then `sycl::future::get` would return the value associated with the invocation of the function. Implementation-wise `sycl::future` would have a reference to the `task_sequence` and an index into the stack of return values so it can look up the right value.

This would change the semantics a little from what is proposed here but I think it would align even closer with `std::async`.

Reply:
I don't think this is meant to align very closely/exactly with std::async, as we do not currently want to give users the ability to call get() out-of-order relative to the order that they called async, as that may required complicated synchronization hardware to block on get calls until the correct result is available, and then kind of short-circuit that result out of the results data structure so it gets past any potential completedinvocations that were launched before it (in async order) and are waiting in the results data structure.